shell/server/proxy.js

```diff
@@ -310,14 +319,25 @@ function openSessionInternal(grainId, user, title, apiToken) {
 
   updateLastActive(grainId, userId);
 
-  var isOwner = grainInfo.owner === userId;
+  var sessionId = generateSessionId(grainId, userId, cachedSalt);
+  var session = Sessions.findOne({_id: sessionId});
+  if (session) {
+    // TODO(someday): also do some more checks for anonymous sessions (sessions without a userId).
+    if ((session.userId && session.userId !== userId) ||
+        (session.grainId !== grainId)) {
+      cachedSalt = Random.id(22);
```

Comment:
It looks like this branch can only run if we've found a collision in SHA-256, which is a major cryptographic event. Either that, or we have a serious bug. I think you should log and throw an exception rather than try a new salt.